examples/mir_demo_server/server_example_decoration.cpp

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright © 2019 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: Alan Griffiths <alan@octopull.co.uk>
+ */
+
+#include "server_example_decoration.h"
+#include "server-decoration_wrapper.h"
+#include <miral/window.h>
+#include <future>
+#include <thread>
+
+namespace
+{
+struct ServerDecoration :
+    mir::wayland::ServerDecoration
+{
+    ServerDecoration(struct wl_client *client, struct wl_resource *parent, uint32_t id, wl_resource *surface) :
+        mir::wayland::ServerDecoration::ServerDecoration(client, parent, id),
+        surface{surface}
+    {
+        send_mode_event(decoration_mode);
+        if (auto window = miral::window_for(client, surface))
+            puts("********************** window not null **********************");
+        else
+            puts("************************ window NULL ************************");
+    }
+
+    ~ServerDecoration() override
+    {
+        test.join();
+    }
+
+    void release() override
+    {
+        destroy_wayland_object();
+    }
+
+    void request_mode(uint32_t mode) override
+    {
+        if (mode != decoration_mode)
+            decoration_mode = mode;
+
+        send_mode_event(decoration_mode);
+    }
+
+    std::thread test{[this]
+        {
+            // We put this on a thread and sleep because we need to wait for the client
+            // to 'commit' before the Window object exists
```

Comment:
This seems like it'll be a bit of a sharp edge - particularly if extensions want to participate in window setup, like `xdg_decoration`, where the client will want to render their first committed buffer differently depending on what the shell sends.

Do we have any plan of how to address this? Will we allow `Window`s that aren't visible (due to not having an associated buffer yet)?

Reply:
This is one of the things I've been trying resolve, I don't have a perfect, general solution.

At the root is the different initialization paradigm adopted by the mirclient API and Wayland protocols. The former has a single state initialization and sends all the setup information to the server at once, the latter has a multi-stage initialization: create an object, populate it, and only then send `commit` for the server to act on it.

We do allow `Window`s that don't have a buffer, but we don't create them from the Wayland frontend as we don't otherwise know when the "populate it" stage of initialization completes. (When we added Wayland support we had bugs caused by code that didn't expect new Windows to already have a buffer.) 

It seems better to expose what we can do simply than hope for a swift rework of this logic.